src/Voron/Data/Graphs/Hnsw.Parallel.cs

```diff
@@ -0,0 +1,582 @@
+                    return;
+
+                ref Node startingNode = ref _searchState.Nodes[0];
+                Span<byte> span = startingNode.Encode(ref byteBuffer);
+                entryPointNode = Container.Allocate(_searchState.Llt, _searchState.Options.Container, span.Length, out Span<byte> allocated);
+                span.CopyTo(allocated);
+                _searchState.RegisterNodeLocation(EntryPointId, entryPointNode);
+            }
+
+            // Run 1..MaxConcurrentBatches batches here, depending on how much work we have to run
+            int numberOfBatches = Math.Max(1, _searchState.CreatedNodesCount / MaxConcurrentBatches);
+            // but not too much...
+            int maxTasks = Math.Min(numberOfBatches, MaxConcurrentBatches);
+            NodePlacementScheduler scheduler = new(this, maxTasks);
+            scheduler.Run();
+        }
+
+        private class NodePlacement(Registration parent, NodePlacementScheduler scheduler)
+        {
+            private readonly SearchState _searchState = parent._searchState;
+            private readonly List<int> _candidates = [];
+            private readonly List<int> _nearestIndexes = [];
+            private readonly List<int> _indexes = [];
+            private readonly List<int> _requiresEdgeFiltering = [];
+            private readonly List<UnmanagedSpan> _vectors = [];
+            private readonly HashSet<int> _visited = [];
+            private readonly PriorityQueue<int, float> _candidatesQ = new();
+            private readonly PriorityQueue<int, float> _nearestEdgesQ = new();
+
+            public async Task RunAsync()
+            {
+                while (true)
+                {
+                    var currentNodeIndex = parent._nextNodeIndex++;
+                    if (currentNodeIndex >= _searchState.CreatedNodesCount)
+                        break;
+
+                    await FindGraphPlacementForNode(currentNodeIndex);
+                }
+            }
+
+            private async Task FindGraphPlacementForNode(int currentNodeIndex)
+            {
+                var currentMaxLevel = _searchState.Options.CurrentMaxLevel(_searchState.CreatedNodesCount - currentNodeIndex);
+                int nodeRandomLevel = GetLevelForNewNode(currentMaxLevel);
+                UnmanagedSpan insertedVector;
+                {
+                    //  scoping n here, to avoid "leaking" the reference and async issues
+                    ref var n = ref _searchState.GetNodeByIndex(currentNodeIndex);
+                    n.EdgesPerLevel.SetCapacity(_searchState.Llt.Allocator, nodeRandomLevel + 1);
+                    insertedVector = n.GetVectorUnmanagedSpan(_searchState);
+                }
+                await SearchNearestAcrossLevelsAsync(insertedVector, currentMaxLevel);
+                for (int level = nodeRandomLevel; level >= 0; level--)
+                {
+                    int startingPointIndex = _nearestIndexes[level];
```

Comment:
This algorithm could use some inline comments explaining what it does and the tradeoffs involved.